staging/src/k8s.io/apiserver/pkg/cel/mutation/optional_test.go

```diff
@@ -0,0 +1,134 @@
+*/
+
+package mutation
+
+import (
+	"reflect"
+	"strings"
+	"testing"
+)
+
+// TestCELOptional is an exploration test to demonstrate how CEL optional library
+// behave for the use cases that the mutation library requires.
+func TestCELOptional(t *testing.T) {
+	for _, tc := range []struct {
+		name                 string
+		expression           string
+		expectedValue        any
+		expectedCompileError string
+	}{
+		{
+			// question mark syntax still requires the field to exist in object construction
+			name: "construct non-existing field, compile error",
+			expression: `Object{
+				?nonExisting: optional.none()
+			}`,
+			expectedCompileError: `undefined field 'nonExisting'`,
+		},
+		{
+			// The root cause of the behavior above is that, has on an object (or Message in the Language Def),
+			// still require the field to be declared in the schema.
+			//
+			// Quoting from
+			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
+			//
+			// To test for the presence of a field, the boolean-valued macro has(e.f) can be used.
+			//
+			// 2. If e evaluates to a message and f is not a declared field for the message,
+			// has(e.f) raises a no_such_field error.
+			name:                 "has(Object{}), de-sugared, compile error",
+			expression:           "has(Object{}.nonExisting)",
+			expectedCompileError: `undefined field 'nonExisting'`,
+		},
+		{
+			name: "construct existing field with none, empty object",
+			expression: `Object{
+				?existing: optional.none()
+			}`,
+			expectedValue: map[string]any{
+				// "existing" field was not set.
+			},
+		},
+		{
+			name:          "object of zero value, ofNonZeroValue",
+			expression:    `Object{?spec: optional.ofNonZeroValue(Object.spec{?replicas: Object{}.?replicas})}`,
+			expectedValue: map[string]any{
+				// "spec" field was not set.
+			},
+		},
+		{
+			name:                 "access non-existing field, return none",
+			expression:           `Object{}.?nonExisting`,
+			expectedCompileError: `undefined field 'nonExisting'`,
+		},
+		{
+			name:       "access existing field, return none",
+			expression: `Object{}.?existing`,
+			// types.OptionalNone.Value() is nil
+			expectedValue: nil,
+		},
+		{
+			name:       "map non-existing field, return none",
+			expression: `{"foo": 1}[?"bar"]`,
+			// types.OptionalNone.Value() is nil
+			expectedValue: nil,
+		},
+		{
+			// Map has a different behavior than Object
+			//
+			// Quoting from
+			// https://github.com/google/cel-spec/blob/master/doc/langdef.md#field-selection
+			//
+			// To test for the presence of a field, the boolean-valued macro has(e.f) can be used.
+			//
+			// 1. If e evaluates to a map, then has(e.f) indicates whether the string f is
+			// a key in the map (note that f must syntactically be an identifier).
+			//
+			name: "has on a map, de-sugared, return false",
+			// has marco supports only the dot access syntax.
+			expression:    `has({"foo": 1}.bar)`,
```

Comment:
Consider adding the affirmative tests for presence testing and field selection as well.

Reply:
Done! Thank you.